File: backend/src/main/java/com/example/knowledgehub/backend/resource/LearningResourceDtos.java
package com.example.knowledgehub.backend.resource;

import com.example.knowledgehub.backend.category.Category;
import jakarta.validation.constraints.NotBlank;

public class LearningResourceDtos {

    public record LearningResourceResponse(
            Long id,
            String name,
            String description,
            String url,
            Long categoryId
    ) {
    }

    public record LearningResourceCreateRequest(
            @NotBlank String name,
            String description,
            @NotBlank String url,
            Long categoryId
    ) {
    }

    public record LearningResourceUpdateRequest(
            String name,
            String description,
            String url,
            Long categoryId
    ) {
    }

    public static LearningResourceResponse toResponse(LearningResource entity) {
        if (entity == null) return null;
        return new LearningResourceResponse(
                entity.getId(),
                entity.getName(),
                entity.getDescription(),
                entity.getUrl(),
                entity.getCategory() != null ? entity.getCategory().getId() : null
        );
    }

    public static LearningResource fromCreateRequest(LearningResourceCreateRequest request, Category category) {
        if (request == null) return null;
        return new LearningResource(
                request.name(),
                request.description(),
                request.url(),
                category
        );
    }

    public static void applyUpdate(LearningResource entity, LearningResourceUpdateRequest update, Category category) {
        if (entity == null || update == null) return;
        if (update.name() != null) entity.setName(update.name());
        if (update.description() != null) entity.setDescription(update.description());
        if (update.url() != null) entity.setUrl(update.url());
        if (update.categoryId() != null) entity.setCategory(category);
    }
}

File: backend/src/main/java/com/example/knowledgehub/backend/resource/LearningResourceDtos.kt
package com.example.knowledgehub.backend.resource

import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull

data class LearningResourceResponse(
    val id: Long?,
    val categoryId: Long?,
    val categoryCode: String,
    val title: String,
    val shortDescription: String,
    val resourceUrl: String,
    val iconUrl: String?,
    val orderIndex: Int?,
)

data class LearningResourceCreateRequest(
    @field:NotNull
    val categoryId: Long,

    @field:NotBlank
    val title: String,

    @field:NotBlank
    val shortDescription: String,

    @field:NotBlank
    val resourceUrl: String,

    val iconUrl: String? = null,
    val orderIndex: Int? = null,
)

data class LearningResourceUpdateRequest(
    val categoryId: Long? = null,
    val title: String? = null,
    val shortDescription: String? = null,
    val resourceUrl: String? = null,
    val iconUrl: String? = null,
    val orderIndex: Int? = null,
)

fun LearningResource.toResponse() = LearningResourceResponse(
    id = id,
    categoryId = category.id,
    categoryCode = category.code,
    title = title,
    shortDescription = shortDescription,
    resourceUrl = resourceUrl,
    iconUrl = iconUrl,
    orderIndex = orderIndex,
)
